lib/ecto/changeset.ex

```diff
@@ -174,7 +174,7 @@ defmodule Ecto.Changeset do
                         types: nil | %{atom => Ecto.Type.t}}
 
   @type error :: {String.t, Keyword.t}
-  @type action :: nil | :insert | :update | :delete | :replace
+  @type action :: nil | :insert | :upsert | :update | :delete | :replace
```

Comment:
Oh, I see why you added this. I think we need to consider upsert either an insert or an update. Maybe it would be best to *not* add an `upsert` function in the repository and instead use `insert` and use the `:on_conflict` option to specify the conflict resolution?